### DISCUSSION TITLE
Search by email and documentation in general

### DISCUSSION THREAD

**Discussion by gl2007** [codeigniter4/shield#816] [Q&A] [answered]
Is there documentation on 
1. what all providers exist and what all methods are available in each method? For e.g. this [managing-users](https://codeigniter4.github.io/shield/quickstart/#managing-users) talks about findbyid but if I have to search by email or other columns in auth_identities/users (search by a logged in admin) then?
2. how do I use relationships? ChapGPT shows hasOne/isOneOf methods, which didn't work (not surprised). Do I have to do the usual query or is there lazyloading of referenced objects?

Will be a bit honest here but it's been a struggle to find things around in Shield. Would have been great to just list down all available classes, their methods, properties, ways to use etc. 

Note: just started looking at the [source](https://github.com/codeigniter4/shield/tree/develop/src)

**Comment by datamweb** [comment]
We have tried our best to provide concise but clear documentation to users. I may not have done well in the sections on this. But fortunately, Shield is an open source project and anyone can contribute to its improvement.

Regarding your question, unfortunately, I did not understand some parts. But if you want to search based on email, do the following.

```php
$users = auth()->getProvider();
echo $users->findByCredentials(['email' => 'email@xxx.ir'])?->username;
```

**Reply by gl2007** [reply]
thanks! I will try that. Having it opensource is great.
After reading UserIdentityModel.php, I was going ahead in the way below:

```
        $userIdentities = new UserIdentityModel();
        return $userIdentities->getIdentityBySecret('email_password', $email);
```
Will that work?

Also, do you have thoughts on my second question. Any examples would be great.

**Reply by datamweb** [reply]
> Will that work?

It works, but it doesn't return users table data ('username', 'f_name', ...). So I don't suggest it.

> do you have thoughts on my second question

sorry, in general, I did not understand your question.

**Reply by gl2007** [reply]
thanks @datamweb that helps!

**Comment by kenjis** [comment, marked as answer]
> what all providers exist and what all methods are available in each method? 

There are only one provider "User Provider" in Shield.
https://codeigniter4.github.io/shield/concepts/#user-providers

I recommend you see the source code to know all methods. 
It is well documented with PHPDocs.

>    how do I use relationships? ChapGPT shows hasOne/isOneOf methods, which didn't work (not surprised). Do I have to do the usual query or is there lazyloading of referenced objects?

There is no special feature for relationships.
But User Provider handles email/password [User Identities](https://codeigniter4.github.io/shield/concepts/#user-identities).

**Reply by gl2007** [reply]
thanks @kenjis for the clarifications. Yeah, started looking at source code. I have couple other questions, I will create separate thread.
